lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -533,53 +651,299 @@ crypto_acipher_ecc_shared_secret(struct ecc_keypair *private_key __unused,
+	 */
+	switch (algo) {
+#if defined(CFG_CRYPTO_ECB)
+	case TEE_ALG_AES_ECB_NOPAD:
+		cipher_info = get_cipher_info(algo, 128);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC)
+	case TEE_ALG_AES_CBC_NOPAD:
+		cipher_info = get_cipher_info(algo, 128);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CTR)
+	case TEE_ALG_AES_CTR:
+		cipher_info = get_cipher_info(algo, 128);
+		break;
+#endif
+#if defined(CFG_CRYPTO_XTS)
+	case TEE_ALG_AES_XTS:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+#if defined(CFG_CRYPTO_CTS)
+	case TEE_ALG_AES_CTS:
+		return TEE_ERROR_NOT_SUPPORTED;
+#endif
+#if defined(CFG_CRYPTO_ECB)
+	case TEE_ALG_DES_ECB_NOPAD:
+		cipher_info = get_cipher_info(algo, MBEDTLS_KEY_LENGTH_DES);
+		break;
+	case TEE_ALG_DES3_ECB_NOPAD:
+		cipher_info = get_cipher_info(algo,
+					MBEDTLS_KEY_LENGTH_DES_EDE);
+		break;
+#endif
+#if defined(CFG_CRYPTO_CBC)
+	case TEE_ALG_DES_CBC_NOPAD:
+		cipher_info = get_cipher_info(algo, MBEDTLS_KEY_LENGTH_DES);
+		break;
+	case TEE_ALG_DES3_CBC_NOPAD:
+		cipher_info = get_cipher_info(algo,
+					MBEDTLS_KEY_LENGTH_DES_EDE);
+		break;
+#endif
+	default:
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (!cipher_info)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	res = cipher_get_ctx_size(algo, &ctx_size);
+	if (res)
+		return res;
+
+	ctx = calloc(1, ctx_size);
+	if (!ctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	mbedtls_cipher_init(ctx);
+
+	lmd_res = mbedtls_cipher_setup(ctx, cipher_info);
+	if (lmd_res != 0) {
+		crypto_cipher_free_ctx(ctx, algo);
+		FMSG("mbedtls_cipher_setup failed, res is 0x%x", -lmd_res);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	*ctx_ret = ctx;
+	return TEE_SUCCESS;
 }
 
-void crypto_cipher_copy_state(void *dst_ctx __unused, void *src_ctx __unused,
-			      uint32_t algo __unused)
+void crypto_cipher_free_ctx(void *ctx, uint32_t algo __maybe_unused)
 {
+	size_t ctx_size __maybe_unused;
+
+	/*
+	 * Check that it's a supported algo, or crypto_cipher_alloc_ctx()
+	 * could never have succeded above.
+	 */
+	assert(!cipher_get_ctx_size(algo, &ctx_size));
+	mbedtls_cipher_free(ctx);
+	free(ctx);
 }
 
-TEE_Result crypto_cipher_init(void *ctx __unused, uint32_t algo __unused,
-			      TEE_OperationMode mode __unused,
-			      const uint8_t *key1 __unused,
-			      size_t key1_len __unused,
+void crypto_cipher_copy_state(void *dst_ctx, void *src_ctx,
+				uint32_t algo __unused)
+{
+	mbedtls_cipher_clone(dst_ctx, src_ctx);
+}
+
+TEE_Result crypto_cipher_init(void *ctx, uint32_t algo,
+			      TEE_OperationMode mode,
+			      const uint8_t *key1, size_t key1_len,
 			      const uint8_t *key2 __unused,
 			      size_t key2_len __unused,
-			      const uint8_t *iv __unused,
-			      size_t iv_len __unused)
+			      const uint8_t *iv,
+			      size_t iv_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+	int lmd_res;
+	int lmd_mode;
+
+	if (!ctx)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Now, re-set the cipher info according the really key length. */
```

Comment:
```
/*
 * Now, re-set the cipher info according the real key length.
 * Note that changing key length is safe since the underlaying
 * allocation is the same.
 */

Reply:
Ok, it will be more clearly.